### DISCUSSION TITLE
error when trigger kdump in qcow2 image

### DISCUSSION THREAD

**Discussion by lisalala** [cloud-hypervisor/cloud-hypervisor#4071] [General]
in vm, trigger kdump by: echo c > /proc/sysrq/trigger 
vm hang-task and got a log :cloud-hypervisor: 38.559998964s: <_disk0_q0> ERROR:virtio-devices/src/block.rs:288 -- Failed to process queue (submit): RequestExecuting(AsyncRead(ReadVectored(Os { code: 95, kind: Other, message: "Operation not supported" })))
	
when 'echo c' as above, callstack like this:

```
(gdb) bt
#0  block_util::qcow_sync::QcowSync::new (qcow_file=..., semaphore=Arc(strong=2, weak=0) = {...}) at block_util/src/qcow_sync.rs:50
#1  0x0000555555c80b16 in <block_util::qcow_sync::QcowDiskSync as block_util::async_io::DiskFile>::new_async_io (self=0x7fffe00399e0, _ring_depth=128)
    at block_util/src/qcow_sync.rs:34
#2  0x00005555559d4a34 in <virtio_devices::block::Block as virtio_devices::device::VirtioDevice>::activate (self=0x7fffe0005900, mem=..., 
    interrupt_cb=Arc(strong=2, weak=0) = {...}, queues=Vec(size=0), queue_evts=Vec(size=0)) at virtio-devices/src/block.rs:544
#3  0x0000555555a2b0e6 in virtio_devices::transport::pci_device::VirtioPciDevice::activate (self=0x7fffe004ca40)
    at virtio-devices/src/transport/pci_device.rs:931
#4  0x0000555555a2b332 in virtio_devices::transport::pci_device::VirtioPciDevice::maybe_activate (self=0x7fffe004ca40)
    at virtio-devices/src/transport/pci_device.rs:939
#5  0x0000555555912d26 in vmm::device_manager::DeviceManager::activate_virtio_devices (self=0x7fffe0002c30) at vmm/src/device_manager.rs:3656
#6  0x00005555558b9156 in vmm::vm::Vm::activate_virtio_devices (self=0x7ffff6845b90) at vmm/src/vm.rs:1987
#7  0x000055555571a491 in vmm::Vmm::control_loop (self=0x7ffff6845b58, api_receiver=Arc(strong=1, weak=0) = {...}) at vmm/src/lib.rs:2449
#8  0x00005555556d06e4 in vmm::start_vmm_thread::{{closure}} () at vmm/src/lib.rs:362
    at /builddir/build/BUILD/rustc-1.47.0-src/library/std/src/sys_common/backtrace.rs:137

```

code:
```
block_util/src/qcow_sync.rs:
	impl DiskFile for QcowDiskSync {
	...

    fn new_async_io(&self, _ring_depth: u32) -> DiskFileResult<Box<dyn AsyncIo>> {
        Ok(Box::new(QcowSync::new(
			self.qcow_file.clone(),
            self.semaphore.clone(),
        )) as Box<dyn AsyncIo>)
    }
}
```

**at the moment, self.qcow_file is the object created when vm booting up, so it's members (such as l1_table) is different the qcow file on disk which has been changed.**
So, is it inappropriate to use a clone of self.qcow_file in this place? Or can we reload it's members from disk?

**Comment by sboeuf** [comment]
>at the moment, self.qcow_file is the object created when vm booting up, so it's members (such as l1_table) is different the qcow file on disk which has been changed.
So, is it inappropriate to use a clone of self.qcow_file in this place?

No it's not because we're cloning an `Arc<Mutex<QcowFile>>`, meaning the cloned object points to the same underlying address in memory.

>Or can we reload it's members from disk?

No need to reload anything since the `QcowFile` implementation includes the update of the L1 and L2 tables based on accesses to the file.

**Comment by rbradford** [comment]
You get that error if you have a QCOW file with compressed blocks. From 592a65f52e7fa3076d6c5f5c15ef076103b82f95 you get an error message earlier.

**Reply by rbradford** [reply]
I just tested this with a QCOW file I know doesn't have compressed blocks and I do not see the error.
